celery/app/backends.py

```diff
@@ -57,13 +57,23 @@ def by_name(backend=None, loader=None,
 
 
 def by_url(backend=None, loader=None):
-    """Get backend class by URL."""
+    """Get backend class by URL. Return a class object and the url.
+
+    New in version 4.2.?: url is returned unmodified. Previously any leading
+    schema followed by a plus sign would be stripped off. Now backends that
+    require special formatting must handle that formatting in their own
+    constructors.
+    """
     url = None
     if backend and '://' in backend:
+        # There are two types of uris we'll see:
+        # * <backend>+<backend-defined-ext>:// e.g., mongodb+srv://
+        # * <backend>+<backend-subtype>:// e.g., cache+memory:// (cache is the
+        #   only one tested and documented thus far
         url = backend
         scheme, _, _ = url.partition('://')
         if '+' in scheme:
-            backend, url = url.split('+', 1)
+            backend, _, _ = scheme.partition('+')
         else:
```

Comment:
I made this dumb. Let CacheBackend (and other url munging backends) handle its own munging.

Reply:
This change will break backwards compatibility, especially for the SQLAlchemy backend. Please keep the scope to MongoDB and let's make sure that every change here is backwards compatible.

Reply:
@josephtate you may be right and we need to evaluate this more thoroughly. In order to maintain the scope in this PR, you can open a separate issue & PR to explain and implement your design. What do you think?

Reply:
You know what. Nevermind. I'll put this back. I don't care enough to re-architect this.

Reply:
I've got what I think is a good middle ground. I'll commit it and you can tell me what you think.